File: backend/src/sql/orgs.ts
import { query } from "../db";
import { NotFoundError } from "../errors";

export const SQL_SELECT_ORG_BY_SLUG = `
  SELECT id, slug, name, plan, settings, created_at
  FROM organizations
  WHERE slug = $1
`;

export const SQL_SELECT_ORGS_FOR_USER = `
  SELECT o.id, o.slug, o.name, o.plan, o.settings, o.created_at
  FROM organizations o
  JOIN memberships m ON m.org_id = o.id
  WHERE m.user_id = $1
  ORDER BY o.created_at DESC
`;

export const SQL_INSERT_ORG = `
  INSERT INTO organizations (slug, name, plan)
  VALUES ($1, $2, 'free')
  RETURNING id, slug, name, plan, settings, created_at
`;

export const SQL_INSERT_MEMBERSHIP = `
  INSERT INTO memberships (org_id, user_id, role)
  VALUES ($1, $2, $3)
`;

export const SQL_SELECT_MEMBERS = `
  SELECT m.id, m.org_id, m.user_id, m.role, m.joined_at,
         u.id AS user_id_ref, u.email, u.full_name, u.avatar_url
  FROM memberships m
  LEFT JOIN users u ON u.id = m.user_id
  WHERE m.org_id = $1
`;

export const SQL_DELETE_MEMBERSHIP = `
  DELETE FROM memberships
  WHERE org_id = $1 AND user_id = $2
  RETURNING id
`;

export const SQL_UPDATE_MEMBERSHIP_ROLE = `
  UPDATE memberships
  SET role = $3
  WHERE org_id = $1 AND user_id = $2
  RETURNING id, org_id, user_id, role, joined_at
`;

export const SQL_SELECT_USER_BY_ID = `
  SELECT id, email, full_name, avatar_url FROM users WHERE id = $1
`;

export const SQL_SELECT_PENDING_INVITATIONS = `
  SELECT id, org_id, invited_by, email, role, token, accepted_at, expires_at, created_at
  FROM invitations
  WHERE org_id = $1 AND accepted_at IS NULL
`;

export const SQL_CHECK_ORG_SLUG_EXISTS = `SELECT id FROM organizations WHERE slug = $1`;

export const ORG_RETURN_FIELDS = `id, slug, name, plan, settings, created_at`;

export function buildUpdateOrgSql(sets: string[]): string {
  return `
    UPDATE organizations
    SET ${sets.join(", ")}, updated_at = now()
    WHERE slug = $1
    RETURNING ${ORG_RETURN_FIELDS}
  `;
}

export async function getOrgBySlug(slug: string) {
  const result = await query(SQL_SELECT_ORG_BY_SLUG, [slug]);
  const org = result.rows[0];
  if (!org) throw new NotFoundError("Organization not found");
  return org;
}

File: backend/app/sql/orgs.py
from __future__ import annotations

from sqlalchemy import delete, insert, select, update

from app.db.tables import invitations, memberships, organizations, users


def select_org_by_slug(slug: str):
    return select(
        organizations.c.id,
        organizations.c.slug,
        organizations.c.name,
        organizations.c.plan,
        organizations.c.settings,
        organizations.c.created_at,
    ).where(organizations.c.slug == slug)


def select_orgs_for_user(user_id):
    return (
        select(
            organizations.c.id,
            organizations.c.slug,
            organizations.c.name,
            organizations.c.plan,
            organizations.c.settings,
            organizations.c.created_at,
        )
        .join(memberships, memberships.c.org_id == organizations.c.id)
        .where(memberships.c.user_id == user_id)
        .order_by(organizations.c.created_at.desc())
    )


def insert_org(slug: str, name: str):
    return (
        insert(organizations)
        .values(slug=slug, name=name, plan="free")
        .returning(
            organizations.c.id,
            organizations.c.slug,
            organizations.c.name,
            organizations.c.plan,
            organizations.c.settings,
            organizations.c.created_at,
        )
    )


def insert_membership(org_id, user_id, role: str):
    return insert(memberships).values(org_id=org_id, user_id=user_id, role=role)


def select_members(org_id):
    return (
        select(
            memberships.c.id,
            memberships.c.org_id,
            memberships.c.user_id,
            memberships.c.role,
            memberships.c.joined_at,
            users.c.email,
            users.c.full_name,
            users.c.avatar_url,
        )
        .outerjoin(users, users.c.id == memberships.c.user_id)
        .where(memberships.c.org_id == org_id)
    )


def delete_membership(org_id, user_id):
    return (
        delete(memberships)
        .where(memberships.c.org_id == org_id, memberships.c.user_id == user_id)
        .returning(memberships.c.id)
    )


def update_membership_role(org_id, user_id, role: str):
    return (
        update(memberships)
        .where(memberships.c.org_id == org_id, memberships.c.user_id == user_id)
        .values(role=role)
        .returning(
            memberships.c.id,
            memberships.c.org_id,
            memberships.c.user_id,
            memberships.c.role,
            memberships.c.joined_at,
        )
    )


def select_pending_invitations(org_id):
    return select(
        invitations.c.id,
        invitations.c.org_id,
        invitations.c.invited_by,
        invitations.c.email,
        invitations.c.role,
        invitations.c.token,
        invitations.c.accepted_at,
        invitations.c.expires_at,
        invitations.c.created_at,
    ).where(invitations.c.org_id == org_id, invitations.c.accepted_at.is_(None))


def check_org_slug_exists(slug: str):
    return select(organizations.c.id).where(organizations.c.slug == slug)
